src/segment-loader.js

```diff
@@ -717,6 +717,94 @@ export default class SegmentLoader extends videojs.EventTarget {
     };
   }
 
+  abortRequestEarly_(stats) {
+    // If the player is not paused and the current playlist is not the
+    // lowestEnabledRendition, consider the possibility of aborting the current request
+    // early for an emergency downswitch
+    // TODO: Replace timeout with a boolean indicating whether this playlist is the
+    //       lowestEnabledRendition
+    if (!this.hls_.tech_.paused() && this.xhrOptions_.timeout) {
+      const firstByteReceived = Date.now() - stats.firstByteReceived;
+
+      // Wait at least 1 second before using the calculated bandwidth from the
+      // progress event to allow the bitrate to stableize
+      if (firstByteReceived > 1000) {
+        const realBandwidth = stats.bandwidth;
+        const estimatedSize =
+          this.pendingSegment_.duration * this.playlist_.attributes.BANDWIDTH;
+        const currentTime = this.currentTime_();
+        const estimatedDelay =
+          (estimatedSize - (stats.bytesReceived * 8)) / realBandwidth;
+        const buffered = this.buffered_();
+        const bufferedEnd = buffered.length ? buffered.end(buffered.length - 1) : 0;
+        const rebufferingDelay =
+          (bufferedEnd - currentTime) / this.hls_.tech_.playbackRate();
+
+        // consider aborting early if the estimated time to finish the download
+        // is larger than the estimated time until the player runs out of forward buffer
+        if (estimatedDelay > rebufferingDelay) {
+          // get a list of playlists with a lower bandwidth than the current
+          const playlists = this.hls_.playlists.master.playlists.slice().filter(
+            (media) => {
+              return media.attributes &&
+                     media.attributes.BANDWIDTH &&
+                     media.attributes.BANDWIDTH <= this.playlist_.attributes.BANDWIDTH;
+            }
+          ).sort((a, b) => a.attributes.BANDWIDTH - b.attributes.BANDWIDTH);
+
+          let newEstimatedDelay;
+          let safeBandwidth;
+
+          // Find the playlist with the highest bandwidth that the player can load from
+          // safely without needing to rebuffer
+          for (let i = playlists.length - 1; i >= 0; i--) {
```

Comment:
Instead of doing the logic here, I think we can move this over to `playlist-selectors` as a function that takes in the playlists, bandwidth, and forward buffer, and returns the playlist that should be able to provide a segment within the remaining buffer time. It can also make use of some of the player size information to ensure that we over select for our window size.